Enforce consumes interface in output modules

wmtan opened the pull request:
Output modules currently have direct access to EventPrincipal, RunPrincipal, and LuminosityBlockPrincipal. Calls to getByLabel() or getForOutput() functions of Principal bypass the consumes interface, so missing consumes calls will not be detected. This PR modifies output modules so that they do not access Principals directly. Rather, they use new classes EventForOutput, LuminosityBlockForOutput, and RunForOutput, similar to the Event, LuminosityBlock, and Run classes used by other modules. Furthermore, reading data is done only by getByToken(), so the consumes interface is enforced everywhere. getByToken() is the only get method supported by the new classes.

wmtan committed:
Enforce consumes interface in output modules

cmsbuild commented:
A new Pull Request was created by  @wmtan  for CMSSW_8_1_X.

It involves the following packages:

DQMServices/FwkIO
DQMServices/StreamerIO
DataFormats/Common
DataFormats/Provenance
EventFilter/Utilities
FWCore/Framework
FWCore/Modules
IOPool/Input
IOPool/Output
IOPool/Streamer
Mixing/Base


@smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald, @barvic, @wddgit this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 2134fd54d4513d19add2eb95bf719c733fffb14a
I found errors in the following unit tests: 
 
 ---> test runtestRecoLocalCaloHGCalRecProducers had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
---> test runtestSimCalorimetryHGCalSimProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11719/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11719/summary.html

wmtan commented:
@cmsbuild The three failing unit tests fail in the base IB. The failures have nothing to do with this PR.

deguio commented:
please test

deguio commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Framework/interface/EventForOutput.h:
```diff
@@ -0,0 +1,171 @@
+    // Principal class.
+    typedef std::set<BranchID> BranchIDSet;
+    mutable BranchIDSet gotBranchIDs_;
+    void addToGotBranchIDs(Provenance const& prov) const;
```
We don't need to record the provenance for output modules.

wmtan replied:
@Dr15Jones Yes, you are correct. This will be removed.

Dr15Jones commented on FWCore/Framework/interface/TriggerResultsBasedEventSelector.h:
```diff
@@ -29,6 +31,13 @@ namespace edm {
     public:
       NamedEventSelector(std::string const& n, EventSelector const& s) :
```
Is this constructor even needed anymore?

wmtan replied:
@Dr15Jones No, it should not be needed. However, the optimal fix to get rid of it invlves other changes, and will be done in a separate PR, to avoid having to rebuild > 800 packages when testing.

Dr15Jones commented on FWCore/Framework/src/EventForOutput.cc:
```diff
@@ -0,0 +1,112 @@
+    if (result.failedToGet()) {
+      return false;
+    }
+    addToGotBranchIDs(*result.provenance());
```
We shouldn't need to do.

Dr15Jones commented on IOPool/Output/src/RootOutputFile.cc:
```diff
@@ -761,10 +756,74 @@ namespace edm {
     if(productProvenanceVecPtr != nullptr) productProvenanceVecPtr->clear();
   }
 
+  void RootOutputFile::fillBranches(LuminosityBlockForOutput const& lumi) {
```
In order to share the code, would it be better to either make this function templated or to create a base class for all `*ForOutput` classes?

wmtan replied:
@Dr15Jones I agree.  I will create a base class.

Dr15Jones commented:
Done with my comments

wmtan committed:
Output modules do not use gotBranchIDs_

cmsbuild commented:
Pull request #13600 was updated. @smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

wmtan commented:
Don't do anything with this PR yet. There is another update coming.

cmsbuild commented:
-1 
 
Tested at: 2134fd54d4513d19add2eb95bf719c733fffb14a
I found errors in the following unit tests: 
 
 ---> test runtestTqafTopEventSelection had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11755/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11755/summary.html

wmtan committed:
Fix setComplete() call

wmtan committed:
Add base class and fix fillBranches()

slava77 commented:
@wmtan is that update still coming soon?

cmsbuild commented:
Pull request #13600 was updated. @smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

wmtan commented:
@slava77 Coincidentally, the update came two minutes after your query.

wmtan commented:
@Dr15Jones This extensive update addresses all of your comments, except for the one about eliminating the old NamedEventSelector ctor, which we agreed I would do in a separate PR. The update also fixes a pre-existing bug in SubProcess.cc (missing setComplete() calls) that was discovered while testing this update.

wmtan commented:
@Dr15Jones @slava77 
Hold it again. I just noticed that two new files that I added in my latest commit do not appear in this PR, even though they show as committed in my working node.
So, there is some git wierdness going on that I do not fully understand yet.
I will have to push again after this is fixed.

wmtan committed:
New Base Class

cmsbuild commented:
Pull request #13600 was updated. @smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

cmsbuild commented:
Pull request #13600 was updated. @smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

wmtan commented:
@Dr15Jones @slava77 Should be OK now. I had forgotten the "git add" for two new files.

wmtan commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

wmtan commented:
@Dr15Jones I extensively rewrote the RootOutputFile::fillBranches() function, not only to use the new base class for the "ForOutput" class, but to make the function more concise and understandable. I think it is a great improvement.

cmsbuild commented:
-1 
 
Tested at: 680e78fac85177d29efee84c396e2bc06dd7e9c8
I found errors in the following unit tests: 
 
 ---> test TestRunnerPhysicsToolsCondLiteIO had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11801/summary.html

wmtan commented:
@cmsbuild @Dr15Jones The test failures appear to have nothing to do with this PR. The runtestTqafTopEventSelection test fails in the base IB. The TestRunnerPhysicsToolsCondLiteIO test runs successfully in my work area with this PR, although I am running on a slightly older IB.
Also, the error in TestRunnerPhysicsToolsCondLiteIO appears to have no connection to anything I changed. I will investigate further, however, just to be safe.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11801/summary.html

wmtan commented:
@cmsbuild @Dr15Jones OK, I ran the failing TestRunnerPhysicsToolsCondLiteIO with this PR against the latest IB, CMSSW_8_1_X_2016-03-09-1100, and the test passed. I cannot reproduce this test failure. So neither unit test failure is related to this PR.

Dr15Jones commented on IOPool/Output/src/RootOutputFile.cc:
```diff
@@ -682,70 +684,53 @@ namespace edm {
-      }
-    }
-    
+    bool const fastCloning = (branchType == InEvent) && (whyNotFastClonable_ == FileBlock::CanFastClone);
```
We can only fast clone events?

Dr15Jones replied:
Ah, I see that was in the original logic.

wmtan replied:
@Dr15Jones A decision was made a long time ago to never fast copy anything in the run or lumi trees. It makes life a lot simpler.

Dr15Jones commented on IOPool/Output/src/RootOutputFile.cc:
```diff
@@ -682,70 +684,53 @@ namespace edm {
-                            (om_->dropMetaData() == PoolOutputModule::DropPrior && produced));
-      bool getProd = (produced || !fastCloning ||
-         treePointers_[branchType]->uncloned(item.branchDescription_->branchName()));
+      bool getProd = (produced || !fastCloning || treePointers_[InEvent]->uncloned(item.branchDescription_->branchName()));
```
There is no guarantee that the item you are looking at is InEvent.

wmtan replied:
@Dr15Jones Actually, there currently is such a guarantee, because if it is not InEvent, fastCloning will always be false and we never get there. However, the guarantee is based on an assumption that may not always be true in the future. So, the code would be more maintainable and more future proof if InEvent were replaced by branchType here, so I will do so,

Dr15Jones commented on IOPool/Output/src/RootOutputFile.cc:
```diff
@@ -682,70 +684,53 @@ namespace edm {
-          productProvenance = provRetriever->branchIDToProvenance(id);
+        bool found = occurrence.getByToken(item.token_, item.branchDescription_->unwrappedTypeID(), result);
+        product = result.wrapper();
+        if(found && keepProvenance) {
```
Couldn't the product not be found but the provenance still be there? This would be the case where we did not write out the product but kept the provenance in a previous step.

Dr15Jones replied:
Ah, I see later that there is a separate check which handles that case.

wmtan replied:
@Dr15Jones Yes, it's the same separate check as before, but it got moved down a few lines, so you hadn't gotten there yet.

Dr15Jones commented:
Done with next review

wmtan committed:
Don't lose provenance

wmtan committed:
Avoid confusion

cmsbuild commented:
Pull request #13600 was updated. @smuzaffar, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @mdhildreth, @cmsbuild, @deguio, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

wmtan commented:
Incorporated latest comments from Chris Jones. Should be all done now.

wmtan commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 1abdfce63189527b70b5106276bb6a5d00ba57a0
I found errors in the following unit tests: 
 
 ---> test runtestTqafTopEventSelection had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11810/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13600/11810/summary.html
